[WIP] sql: record statement timings in system table

couchand opened the pull request:
This adds a new system table `statement_executions`, which stores
the raw statement timings that are aggregated into statement statistics.

Release note (sql change): Record statement timings in a new
system table: `system.statement_executions`.

couchand committed:
sql: record statement timings in system table

This adds a new system table `statement_executions`, which stores
the raw statement timings that are aggregated into statement statistics.

Release note (sql change): Record statement timings in a new
system table: `system.statement_executions`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/33143)
<!-- Reviewable:end -->

couchand commented:
There are certainly a number of things left, including:
- logic tests
- performance benchmarking
- data cleanup
- data retention settings
Putting this up here so it's not just on my laptop while I'm out, and maybe I can get some feedback too.

couchand committed:
sql: expose statement_logging_enabled session var

Release note: None

vilterp commented:
Exciting! Looking forward to seeing what kind of perf impact and storage usage this has, e.g. when running TPCC.

couchand committed:
sql: flush statement execution log in batches

To avoid a per-statement overhead, we flush the statement execution
log to the system table with a background job, recording it to
memory in the meantime.

Release note: None

couchand committed:
lints

couchand committed:
tests

couchand committed:
add statement key

couchand committed:
tests

couchand committed:
more tests

couchand committed:
query for aggregates

couchand committed:
more test fixes

Release note: None

couchand committed:
wip-stats!

couchand committed:
remove dead code

couchand committed:
store transaction id

couchand committed:
more test fixes

couchand committed:
more test fixes

couchand committed:
lints

couchand committed:
fixup system table

couchand committed:
hack test passing

glennfawcett commented:
> This adds a new system table `statement_executions`, which stores
> the raw statement timings that are aggregated into statement statistics.
> 
> Release note (sql change): Record statement timings in a new
> system table: `system.statement_executions`.

Make sure to include performance metrics along with `statement_executions` and timings.  SqlServer, Oracle, Postgres, ... all include essentially the "Active Session History" where there is a hash of the statement id along with an aggregate of the CPU, IO, Network, LockWaits, etc... are stored as events in the [Active_Session_History](https://docs.oracle.com/cd/E11882_01/server.112/e40402/dynviews_1007.htm#REFRN30299) table.   

Finally, once we have the History and can point to a SQL statement, we need to also be storing the Query Execution Plans for long running queries.  Oracle has something called SQLMON which stores the Plan along with the timing of each phase of the query.  This is REALLY useful to pinpoint an exact query that is causing issues.

This presentation from [Tanel Ponder](http://www.nocoug.org/download/2016-02/Poder_Real_Time_SQL_Monitoring_in_Oracle_Database_12c.pdf) does a great job of describing the process.
